src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/PemEncoding.cs

```diff
@@ -66,33 +93,68 @@ public static PemFields Find(ReadOnlySpan<char> pemData)
         /// always uses lax rules.
         /// </remarks>
         public static bool TryFind(ReadOnlySpan<char> pemData, out PemFields fields)
+        {
+            return TryFindCore<char, Utf16PemEncoder>(pemData, out fields);
+        }
+
+        /// <summary>
+        ///   Attempts to find the first PEM-encoded data.
+        /// </summary>
+        /// <param name="pemData">
+        ///   The text containing the PEM-encoded data.
+        /// </param>
+        /// <param name="fields">
+        ///   When this method returns, contains a value
+        ///   that specifies the location, label, and data location of the encoded data;
+        ///   or that specifies those locations as empty if no PEM-encoded data is found.
+        ///   This parameter is treated as uninitialized.
+        /// </param>
+        /// <returns>
+        ///   <see langword="true" /> if PEM-encoded data was found; otherwise <see langword="false" />.
+        /// </returns>
+        /// <remarks>
+        ///   IETF RFC 7468 permits different decoding rules. This method always uses lax rules.
+        /// </remarks>
+        public static bool TryFindUtf8(ReadOnlySpan<byte> pemData, out PemFields fields)
+        {
+            if (!Utf8.IsValid(pemData))
```

Comment:
I am open to some discussion here. This does mean we essentially do a two-pass of UTF-8 data - once to check validity and again to actually find the PEM.

I think this is okay. We have modern vectorized implementations of this now.

We could instead document it in a remark "Callers are expected to ensure the input is well-formed UTF-8". However I think this is a little accident prone since bytes off the network can be binary or byte data. I don't want the PEM finder to yank a UTF-8 encoded certificate that's embedded in a PE, for example.

Reply:
What happens if this isn't done? Non-UTF8 data still fails, just later? Non-UTF8 data is never found? Something else?

Reply:
The question comes down to, do we care about the things we aren't looking for well formedness.

We are looking for, in UTF-8, some text. It looks like `-----BEGIN `, but what it looks like isn't too important for this. We know the PEM is valid without `Utf8.IsValid(pemData)` because it gets validated later on (the label and the base-64 contents).

But the question is about the stuff around the PEM. Consider an input that looks something like this (escaped characters are for readability):

```
\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF
-----BEGIN CERTIFICATE-----
MII...a bunch of base64 here
-----END CERTIFICATE-----
\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF\xFF
```

Is this input valid UTF-8? No. It's a hunk of UTF-8 text that appears in the middle of a binary blob. Is that valid? Well, "it depends". This is a "confusion" issue because it could, for example, be a Win32 PE that has a certificate embedded in a resource. I would not expect the above to work. The input is not UTF-8 text.

But according to the spec, text _is_ allowed to appear outside of the PEM encapsulation boundary. I would expect this to parse:

```
こんばんは
-----BEGIN CERTIFICATE-----
MII...a bunch of base64 here
-----END CERTIFICATE-----
こんばんは
```

Assuming the input is valid and correctly encoded UTF-8.

Reply:
In the X509CertificateLoader security design document (which I'll get around to putting in the repository one of these days), it says:

> The textual encoding loaders will ignore any text outside the relevant envelope, and as they treat the data as 8-bit ASCII rather than UTF-8, there is no notion of "invalid" data outside the envelope.

I don't immediately see a test that verifies that, but changing LoadWrappingCertificates_PEM_WithSurroundingText to have "\u00FF\u00FF" between seven and years still successfully loads on Windows.

https://github.com/dotnet/runtime/blob/275ea587d87e9eacc36dfd719e85155a1e7014b6/src/libraries/Common/tests/System/Security/Cryptography/X509Certificates/X509CertificateLoaderTests.cs#L329-L340

We can certainly move to a single PEM interpretation and change it from passing to failing; but if we're checking data outside the encapsulation boundaries then using it in X509CertificateLoader would be a breaking change.

We... could... take it back to API Review and say we want to rename Utf8 to Ascii.  Or just doc that data outside encapsulation boundaries isn't checked for validity (basically treating it as if we read malformed data as the replacement character).  Since all UTF-8 "continuation" code units match `0b10xx_xxxx`, the part where we match the Pre-EB as the beginning of the buffer or preceded by a newline means there's no real ambiguity... just an untidiness of reading a PEM blob out of (e.g.) a PE file resources table.

That said, we could just declare it a breaking change in the loader. (Although, if we'd consider that an unacceptable breaking change for X509Certificate2..ctor then we'd still not get maximum utility from the new functions)

Reply:
I am okay dropping it. As long as we made a conscious decision about it.